doc/aespasswd.man
.TH aespasswd 1 "2004 Feb 8"
.\" The first line of this file must contain the '"[e][r][t][v] line
.\" to tell man to run the appropriate filter "t" for table.
.\"
.\"	$Id$
.\"
.\"######################################################################
.\"#									#
.\"#			   Copyright (C)  2004				#
.\"#	     			Internet2				#
.\"#			   All Rights Reserved				#
.\"#									#
.\"######################################################################
.\"
.\"	File:		aespasswd.man
.\"
.\"	Author:		Jeff Boote
.\"			Internet2
.\"
.\"	Date:		Sun Feb 08 00:23:56 MST 2004
.\"
.\"	Description:	
.\"
.SH NAME
aespasswd \- Used to create and manage an AES keyfile.
.SH SYNOPSIS
.B aespasswd 
[\fI-n\fR] [\fI-d\fR] \fI-f keyfile identity
.SH OPTIONS
.TP
\fB\-n\fR
Create the keyfile
.TP
\fB\-d\fR
Delete given identity from keyfile
.TP
\fB\-f keyfile\fR
Specifies file that holds identity/key pairs
.SH DESCRIPTION
.B aespasswd
is used to create and manage files that hold identity/key pairs. It is
primarily used to manage the \fIbwctld.keys\fR file for \fBbwctld\fR
and the \fIowampd.keys\fR file for \fBowampd\fR.
.PP
If the \fI\-d\fR option is not specified, then \fBaespasswd\fR prompts
the caller for a passphrase. The passphrase is hashed using an internal
MD5 algorithm to generate a key that is then saved in the \fIkeyfile\fR
associated with the given \fIidentity\fR. If the given \fIidentity\fR
already exists in the \fIkeyfile\fR, the previous key is overwritten with
the new one.
.PP
\fIkeyfiles\fR generated by \fBaespasswd\fR are formatted for use with
\fBBWCTL\fR and \fBOWAMP\fR.
.SH KEYFILE FORMAT
\fBaespasswd\fR generates lines of the format:
.LP
test	54b0c58c7ce9f2a8b551351102ee0938
.PP
An \fIidentity\fR, followed by whitespace, followed by a hex encoded 128\-bit
number, that is suitable to be used as a symmetric AES key.
.PP
No other text is allowed on these lines; however, comment lines may be
added. Comment lines are any line where the first non-white space character
is '\fB#\fR'.
.SH EXAMPLES
\fBaespasswd \-f /usr/local/etc/bwctld.keys testuser\fP
.IP
Adds a key for the identity \fItestuser\fR. The user is prompted for
a passphrase. If the file does not exist, an error message will
be printed and no action will be taken.
.LP
\fBaespasswd \-f /usr/local/etc/bwctld.keys \-n testuser\fP
.IP
Creates the file before doing the same as above. If the file already
exists, an error message will be printed and no action will be taken.
.LP
\fBaespasswd \-f /usr/local/etc/bwctld.keys \-d testuser\fP
.IP
Deletes the identity \fItestuser\fR from the keyfile.
If the file does not exist, an error message will be printed and no action will be taken.
.LP
.SH SECURITY CONSIDERATIONS
The keys in the \fIkeyfile\fR are not encrypted in any way. The security
of these keys is completely dependent upon the security of the system and the
discretion of the system administrator.
.SH RESTRICTIONS
\fIidentity\fR names are restricted to 16 characters, and passphrases
are limited to 1024 characters.
.SH SEE ALSO
owping(1), owampd(1), bwctl(1), bwctld(1)
and the \%http://e2epi.internet2.edu/owamp and
\%http://e2epi.internet2.edu/bwctl web sites.
.SH ACKNOWLEDGMENTS
This material is based in part on work supported by the National Science
Foundation (NSF) under Grant No. ANI-0314723. Any opinions, findings and
conclusions or recommendations expressed in this material are those of
the author(s) and do not necessarily reflect the views of the NSF.


debian/changelog
i2util (1.2.2-2) unstable; urgency=low
 
  * to be filled in

 -- Antoine Delvaux <antoine.delvaux@dante.net>  Thu, 19 Mar 2015 11:18:10 +0000

i2util (1.2.2-1) unstable; urgency=low

  * Adding gbp.conf for git-buildpackage
  * Calling bootstrap.sh before launching the build process

 -- Antoine Delvaux <antoine.delvaux@dante.net>  Tue, 10 Mar 2015 19:46:03 +0000

i2util (1.2-1) unstable; urgency=low

  * Initial release (Closes: #753894)

 -- Raoul Gunnar Borenius <borenius@dfn.de>  Mon, 06 Oct 2014 08:24:01 +0200


aespasswd/aespasswd.c
/*
 *      $Id$
 */
/************************************************************************
*									*
*			     Copyright (C)  2003			*
*				Internet2				*
*			     All Rights Reserved			*
*									*
************************************************************************/
/*
 *	File:		aespasswd.c
 *
 *	Author:		Jeff W. Boote
 *			Internet2
 *
 *	Date:		Thu Dec 18 14:41:59 MST 2003
 *
 *	Description:	
 *
 *    License:
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <limits.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <ctype.h>
#include <libgen.h>
#include <assert.h>
#include <I2util/util.h>

#if defined HAVE_DECL_OPTRESET && !HAVE_DECL_OPTRESET
int optreset;
#endif

static void
usage(
	const char	*progname,
	const char	*msg
	)
{
	if(msg) fprintf(stderr, "%s: %s\n", progname, msg);
	fprintf(stderr,"usage: %s %s\n",
			progname,
			"[arguments] identity"
			);
	fprintf(stderr,"\n");
	fprintf(stderr,
"   -f fname       filename to edit [required option]\n"
"   -n             create keyfile\n"
"   -d             delete given identity from file\n"
	);

	return;
}

static int	do_create = 0;
static int	do_delete = 0;
static char	*keyfname;
static char	*idname;

int
main(
		int	argc,
		char	**argv
    )
{
    char            *progname;
    static char     *copts = "hf:nd";
    int             ch;
    int             lockfd;
    char            lockfname[PATH_MAX];
    char            *dname;
    size_t          len;
    struct flock    flk;
    FILE            *fromfp;
    FILE            *tofp;
    uint8_t         aeskey[I2KEYLEN];
    char            *lbuf=NULL;
    size_t          lbuf_max=0;
    int             rc;

    if((progname = strrchr(argv[0],'/'))){
        progname++;
    }else{
        progname = *argv;
    }

    opterr = 0;
    while((ch = getopt(argc,argv,copts)) != -1){
        switch (ch) {
            case 'f':
                if(!(keyfname = strdup(optarg))){
                    fprintf(stderr,"%s:%d: malloc(): %s\n",
                            __FILE__,__LINE__,
                            strerror(errno));
                    exit(errno);
                }
                break;
            case 'n':
                do_create++;
                break;
            case 'd':
                do_delete++;
                break;
            case 'h':
            case '?':
            default:
                usage(progname,"");
                exit(0);
                /* UNREACHED */
        }
    }
    argc -= optind;
    argv += optind;

    if(!keyfname){
        usage(progname,"-f option is required");
        exit(1);
    }

    if(argc != 1){
        usage(progname,NULL);
        exit(1);
    }

#define STR(val)	#val
#define STREXP(val)     STR(val)
    len = strlen(argv[0]);
    if(len > I2MAXIDENTITYLEN){
        usage(progname,"identity can't be more than "
                STREXP(I2MAXIDENTITYLEN)
                " characters");
        exit(1);
    }

    if(!(idname = strdup(argv[0]))){
        fprintf(stderr,"%s:%s:%d: malloc(): %s\n",
                progname,__FILE__,__LINE__,
                strerror(errno));
        exit(errno);
    }

    fromfp = fopen(keyfname,"r+");
    if(do_create && (fromfp != NULL)){
        fprintf(stderr,"%s: -n option specified: %s exists\n\n",
                progname,keyfname);
        usage(progname,NULL);
        exit(1);
    }
    if(!do_create && !fromfp){
        fprintf(stderr,"%s: Can't open \'%s\': %s\nSee -n option?\n\n",
                progname,keyfname,strerror(errno));
        usage(progname,NULL);
        exit(1);
    }

    /* ensure string is nul terminated when strncpy is used */
    memset(lockfname,'\0',sizeof(lockfname));

    /* Don't let the last byte be written over */
    len = sizeof(lockfname) - 1;

    /*
     * copy keyfname to lockfname(used as tmp var here) before calling
     * dirname to ensure it doesn't modify keyfname.
     */
    strncpy(lockfname,keyfname,len);
    if(!(dname = dirname(lockfname))){
        fprintf(stderr,"%s: dirname(): %s\n",progname,strerror(errno));
        exit(1);
    }

    /*
     * what directory are we putting the lockfile in.
     */
    strncpy(lockfname,dname,len);

    /*
     * Concatinate lockfile name on.
     */
    len = sizeof(lockfname) - strlen(lockfname) - 1;
    strncat(lockfname,"/.",len);
    len -= 2; /* strlen("/.") */
    strncat(lockfname,progname,len);
    len -= strlen(progname);
    strncat(lockfname,".lock",len);


    /*
     * open the lock file (also used as the tempfile).
     */
    lockfd = open(lockfname,O_RDWR|O_CREAT,S_IRUSR|S_IWUSR);
    if(lockfd < 0){
        fprintf(stderr,"open(%s): %s\n",lockfname,strerror(errno));
        exit(1);
    }

    /*
     * lock the file
     */
    flk.l_start = 0;
    flk.l_len = 0;
    flk.l_type = F_WRLCK;
    flk.l_whence = SEEK_SET;
    if(fcntl(lockfd,F_SETLK,&flk) < 0){
        fprintf(stderr,"%s: Unable to lock \'%s\': %s\n",
                progname,lockfname,strerror(errno));
        exit(1);
    }

    /*
     * wrap a FILE* around the lockfile
     */
    if(!(tofp = fdopen(lockfd,"r+"))){
        fprintf(stderr,"%s: fdopen(): %s\n",progname,strerror(errno));
        exit(1);
    }

    if(!do_delete){
        char		*passphrase;
        char		ppbuf[1024];
        char		prompt[1024];
        I2MD5_CTX	mdc;
        size_t		pplen;

        rc = snprintf(prompt,sizeof(prompt),
                "Enter passphrase for identity '%s': ",
                idname);
        if((rc < 0) || ((size_t)rc > sizeof(prompt))){
            fprintf(stderr,"%s: Invalid identity '%s'\n",
                    progname,idname);
            exit(1);
        }

        if(!(passphrase = I2ReadPassPhrase(prompt,ppbuf,
                        sizeof(ppbuf),I2RPP_ECHO_OFF))){
            fprintf(stderr,"%s: I2ReadPassPhrase(): %s\n",
                    progname,strerror(errno));
            exit(1);
        }
        pplen = strlen(passphrase);

        I2MD5Init(&mdc);
        I2MD5Update(&mdc,(unsigned char *)passphrase,pplen);
        I2MD5Final((uint8_t *)aeskey,&mdc);
    }

    /*
     * All records that don't match idname will be copied to idname.
     */
    rc = I2ParseKeyFile(NULL,fromfp,0,&lbuf,&lbuf_max,tofp,idname,
            NULL,NULL);
    if(rc < 0){
        fprintf(stderr,"%s: I2ParseKeyFile('%s'): error line %d\n",
                progname,keyfname,rc);
        exit(1);
    }

    if(do_delete){
        /*
         * If we are deleting, the identity should have been found.
         */
        if(!rc){
            fprintf(stderr,"%s: Identity '%s' not found\n",
                    progname,idname);
            exit(1);
        }
    }else{
        /*
         * Not deleting, add the new/changed key line.
         */
        if(I2WriteKeyLine(NULL,tofp,idname,aeskey) < 0){
            fprintf(stderr,"%s:I2WriteKeyLine('%s'): %s\n",
                    progname,keyfname,strerror(errno));
            exit(1);
        }
    }

    /*
     * Copy all remaining records.
     */
    rc = I2ParseKeyFile(NULL,fromfp,rc,&lbuf,&lbuf_max,tofp,idname,
            NULL,NULL);
    if(rc != 0){
        fprintf(stderr,"%s:I2ParseKeyFile('%s'): error line %d\n",
                progname,keyfname,rc);
        exit(1);
    }

    /*
     * Now close the keyfile, and rename the lockfile to the keyfile.
     */
    if(fromfp) fclose(fromfp);
    rename(lockfname,keyfname);
    fclose(tofp);

    exit(0);
}
